### PR TITLE
[Feature] Add rate limiting by IP for tcp/websocket/stats

### PR DESCRIPTION
**Pull request by zquestz**
This adds IP rate limiting. The limits are quite high, and are just there for sanity checks. The rate limit is shared for all connection types from an IP, this means stats calls will count just as much as websocket or tcp calls.

Right now the rate limits are 100 requests per minute for non-tor connections per IP.

Tor connections are rate limited to 500 a minute (they all appear to be the same IP). We might need to adjust this if we see a lot of activity on the tor side.

Let me know what you think. =)

### CODE REVIEW

**Comment by cculianu** [comment]
I think the limit is kind of low, TBH.   What is this trying to prevent exactly? A DDoS?  All the server's stuff is in memory anyway and SYN attacks will kill the server anyway even with this limit in place.

30 connections per minute are already made by the "View pools..." window (it refreshes every 2 seconds).  

On top of that there are something like 7 possible protocol threads that may connect and reconnect a lot especially if they find pools are down.

I think you should set the rate limit much higher or do away with this check entirely. 

Tor will also suffer greatly.

Again -- you are running a server from memory -- no disk access, no DB locks.. nothing.  I do not even think you need a rate limit.   All this does is burn cycles -- DDoS will still succeed because a good DDoS attacks the TCP/IP stack at a lower level before your code is even called.

My two cents.

**Comment by cculianu** [comment]
TL;DR of above:

1. Please define the problem this limit is intended to solve?
2. All I can see it doing is potentially creating a new problem and not solving any existing potential problems.

Thanks.

**Comment by zquestz** [comment]
@cculianu we discussed in Telegram but figure I should put it here. These limits are designed to be above usage, not to restrict normal behavior. Right now it is trivial to open a lot of connections to the server and never send data. This eats file descriptors and we have seen servers crash due to incorrectly set ulimits. Adding basic protections so a single person can't take down the server is absolutely necessary. That being said, the limits are designed to be high, and you are right. We need to make sure they do not interfere with normal behavior.

**Comment by cculianu** [comment]
Ok, sounds good.  Didn't realize it was so easy to take down the server due to kernel ulimit. Fair enough. Ok, so maybe if you can up the limit for regular to 120 -- I don't envision any normal client making more than 2 connections per second.

If you have time -- how is the limit applied? Is it:

-  if any 2 samples exceed the rate (thus if 2 connections come say 200 ms apart, would it break the limit since the instantaneous rate is now 300/minute?)
-  Or is it applied as an average over the period specified (in this case you specified minute as the period of interest). In this case you can do lots of connections in a burst but then have a dead time and it would be ok.

I hope it's the latter :)

**Comment by zquestz** [comment]
It is the later! Also I bumped it to 180. If we have lots of websocket clients they actually do 2 requests to upgrade the websocket connection. This should be way above normal usage thresholds.

**Comment by cculianu** [comment]
Ok -- I googled and looked at the limiter code.  It's pretty useful as a lib for http requests but the rate limiting code itself is pretty simple -- it creates a context on first request and then expires it 1 minute/hour/second whatever later.  And then it just keeps a counter for that context going until you hit the limit.

So as long as within that window you don't exceed the limit -- it's ok.

It's pretty simplistic but what the hey -- it works.

Answered my own question. :)

**Comment by zquestz** [comment]
Yep and for our use case it is ideal. We really didn't want it to restrict burst traffic. =)

**Comment by cculianu** [comment]
(as an aside the algorithm they chose has a peculiarity in that you can theoretically get 2x the limit within a time window if you time it right by putting all your requests at the end of one window and then again another burst at the beginning of the next, heh).

**Comment by zquestz** [comment]
Yeah, you would need to send 1 request, then wait till the end of the window, then blast requests and you could try to eek out 2x. Just not sustained, it would be a burst.

**Comment by cculianu** [comment]
Right.. over time it averages out to the specified rate but you can exceed it temporarily. :)

**Comment by emergent-reasons** [comment]
Code logic LGTM.

I just have one problem that is probably my own ignorance - I can't figure out how it ever uses the option you selected `limiter.WithTrustForwardHeader(true)`. I tried following the code path and didn't find anything.

- `TFH` option gets stored on the limiter itself
- in server loop, `limit.Get(nil, ip)` where IP is a simple IP string as I understand it
- [...] in memory store `limit.Get()`
    - `Increment()` and child calls don't seem to use it.
    - `GetContextFromState()` doesn't seem to use it.

What am I missing? I guess the middleware gets used magically and I am just not seeing it?

**Comment by zquestz** [comment]
> Code logic LGTM.
> 
> I just have one problem that is probably my own ignorance - I can't figure out how it ever uses the option you selected `limiter.WithTrustForwardHeader(true)`. I tried following the code path and didn't find anything.
> 
> * `TFH` option gets stored on the limiter itself
> * in server loop, `limit.Get(nil, ip)` where IP is a simple IP string as I understand it
> * [...] in memory store `limit.Get()`
>   
>   * `Increment()` and child calls don't seem to use it.
>   * `GetContextFromState()` doesn't seem to use it.
> 
> What am I missing? I guess the middleware gets used magically and I am just not seeing it?

Yep, it doesn't get used in the TCP flow, as this is looking at HTTP headers to determine the real IP. This only works for the http/websocket endpoints where it looks for the X-Forwarded-For header!

https://github.com/ulule/limiter/blob/master/network.go#L34
